securedrop/tests/functional/pageslayout/accessibility.py

```diff
@@ -0,0 +1,191 @@
+import os
+from dataclasses import dataclass
+from enum import Enum
+from pathlib import Path
+from typing import Dict, List, Union
+
+from selenium.webdriver.firefox.webdriver import WebDriver
+
+_ACCESSIBILITY_DIR = (Path(__file__).parent / "accessibility-info").absolute()
+
+_HTMLCS_RUNNER_CODE = """
+var all_messages = [];
+console.log = msg => {
+    all_messages.push(msg);
+};
+
+HTMLCS_RUNNER.run('WCAG2AAA');
+return all_messages;
+"""
+
+
+class MessageType(Enum):
+    ERROR = 1
+    WARNING = 2
+    NOTICE = 3
+
+
+@dataclass
+class Message:
+    """Contains all of the information in a message emitted by HTML CodeSniffer."""
+
+    principle_id: str
+    message: str
+    message_type: MessageType
+    responsible_html: str
+    selector: str
+    element_type: str
+
+    @staticmethod
+    def from_output(output: str) -> "Message":
+        """Parses the output of htmlcs and returns an instance containing all data.
+
+        No processing is performed for flexibility.
+
+        Example message, post-split (note: contents of index 4 contains no newlines, but I had to
+        split it to keep the linter happy):
+
+        0: [HTMLCS] Error
+        1: WCAG2AAA.Principle1.Guideline1_3.1_3_1_AAA.G141
+        2: h2
+        3: #security-level-heading
+        4: The heading structure is not logically nested. This h2 element appears to be the
+           primary document heading, so should be an h1 element.
+        5: <h2 id="security-level-heading" hidden="">...</h2>
+        """
+
+        fields = output.split("|")
+
+        if "Error" in fields[0]:
+            message_type = MessageType.ERROR
+        elif "Warning" in fields[0]:
+            message_type = MessageType.WARNING
+        elif "Notice" in fields[0]:
+            message_type = MessageType.NOTICE
+
+        return Message(
+            message_type=message_type,
+            principle_id=fields[1],
+            element_type=fields[2],
+            selector=fields[3],
+            message=fields[4],
+            responsible_html=fields[5],
+        )
+
+    def __format__(self, _spec: str) -> str:
+        newline = "\n"
+        return f"""
+{self.message_type}: {self.principle_id}
+    {self.message}
+
+    html:
+        {self.responsible_html.replace(newline, f"{newline}        ")}
+        """
+
+
+def sniff_accessibility_issues(driver: WebDriver, locale: str, test_name: str) -> None:
+    """Runs accessibility sniffs on the driver's current page.
+
+    This function is responsible for injecting HTML CodeSniffer into the current page and writing
+    the results to a file. This way, test functions can focus on the setup required to navigate to
+    a particular URL (for example, logging in to get to the messages page).
+    """
+
+    # 1. Retrieve/compute required data
+    with open(f"/usr/local/lib/node_modules/html_codesniffer/build/HTMLCS.js") as htmlcs:
```

Comment:
With the change from `npm install --global` to `npm ci`, this will need to be a relative rather than an absolute system path.

Reply:
No longer applicable, per https://github.com/freedomofpress/securedrop/pull/6745#issuecomment-1503797098